### PR TITLE
Issue #4091: False negative on multiblock tokens with ALONE_OR_SINGLINE option

### PR DESCRIPTION
**Pull request by Vladlis**
#4091 

Report: https://vladlis.github.io/4091/

### CODE REVIEW

**Comment by romani** [comment]
@Vladlis , wercker build is strange , it is problem on hibernate , please try to relaunch wercker build.
If still failed - we can update wercker.yml to checkout to last release code to make build stable.
One more idea is to add "-DskipITs=true" to wercker if some integration tests are trying to run .....

**Comment by romani** [comment]
@Vladlis , should be fixed by https://github.com/checkstyle/checkstyle/commit/a70fd44910afb570efe5b8f40902f78bf8674608 , please rebase.

Teamcity is restarted too.

**Comment by Vladlis** [comment]
@romani 
Rebased.
Report shows no regression: https://vladlis.github.io/4091/.

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/blocks/InputRightCurlyAloneOrSingleline.java [hunk 1]
```diff
@@ -176,4 +176,15 @@ private void foo29() {
...
+
+    void foo30() {
+        if (true) {
+            getClass();} // violation
+
+        for (int i = 0; i == 0; i++) {
+            getClass();} // violation
+
+        while (true) {
+            getClass();}
```
Comment (romani) [inline src/test/resources/com/puppycrawl/tools/checkstyle/checks/blocks/InputRightCurlyAloneOrSingleline.java]: violation ?

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/blocks/RightCurlyCheck.java [hunk 2]
```diff
@@ -245,16 +237,20 @@ private static boolean shouldBeOnSameLine(RightCurlyOption bracePolicy, Details
      * @param bracePolicy option for placing the right curly brace
      * @param details Details for validation
      * @return true if a right curly should be alone on a line.
+     * @noinspection OverlyComplexBooleanExpression
      */
+    // -@cs[BooleanExpressionComplexity] keeping logic in single method makes it clearer
```
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/blocks/RightCurlyCheck.java]: this is not ok..... i will think about this .

Did you consider a split ?

Comment (Vladlis) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/blocks/RightCurlyCheck.java]: I did think about it, but
1) it is actually better to see all the logic in one place - it becomes clearer
2) the best option of splitting I could imagine was smth like:
```java
return shouldBeAloneOnLineWithAloneOption(details)
               && bracePolicy == RightCurlyOption.ALONE
           || shouldBeAloneOnLineWithAloneOrSingleLineOption(details)
               && bracePolicy == RightCurlyOption.ALONE_OR_SINGLELINE
           || details.shouldCheckLastRcurly
               && details.rcurly.getLineNo() == details.nextToken.getLineNo();
```

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/blocks/InputRightCurlyAloneOrSingleline.java (hunk 1 above)
Comment (Vladlis) [inline src/test/resources/com/puppycrawl/tools/checkstyle/checks/blocks/InputRightCurlyAloneOrSingleline.java]: yes, fixed

**Comment by Vladlis** [comment]
@romani ,
this is not a false-positive https://vladlis.github.io/4091/openjdk7/index.html#A11
please look attentively, this line has right curly at column 113

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/blocks/RightCurlyCheck.java (hunk 2 above)
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/blocks/RightCurlyCheck.java]: `&& bracePolicy == .... ` please move to the front and it become good.
Please apply this format of code.

> it is actually better to see all the logic in one place - it becomes clearer

:) yes, it is the always a reason when smb write a code, but it is disaster when you try to read it and understand

Comment (Vladlis) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/blocks/RightCurlyCheck.java]: ok, done

**Comment by Vladlis** [comment]
@romani , please review changes so I could proceed with https://github.com/checkstyle/checkstyle/pull/3969

**Comment by romani** [comment]
@rnveach , please review this PR.

**Pull request merged**